symengine/ruby/ext/symengine/symengine_macros.c

```diff
@@ -0,0 +1,51 @@
+#include "symengine_macros.h"
+#include "symengine.h"
+
+void sympify(VALUE operand2, basic_struct **cbasic_operand2) {
+    basic_struct *num_basic, *den_basic;
+    VALUE new_operand2, num, den;
+
+    switch(TYPE(operand2)) {
+        case T_FIXNUM:
+        case T_BIGNUM:
+            new_operand2 = cbasic_alloc(c_integer);
+            Data_Get_Struct(new_operand2, basic_struct, *cbasic_operand2);
+            GET_SYMINTFROMVAL(operand2, *cbasic_operand2);
+            break;
+
+        case T_RATIONAL:
+            num = rb_funcall(operand2, rb_intern("numerator"), 0, NULL);
+            den = rb_funcall(operand2, rb_intern("denominator"), 0, NULL);
+
+            num_basic = basic_new_heap();
+            den_basic = basic_new_heap();
+
+            GET_SYMINTFROMVAL(num, num_basic);
+            GET_SYMINTFROMVAL(den, den_basic);
+
+            new_operand2 = cbasic_alloc(c_rational);
```

Comment:
In all the other cases, Ruby is responsible to free the `VALUE operand2`, in other words, the caller should not free `cbasic_operand2`, is that your understanding? 

So in that case, the `T_RATIONAL` Ruby type cannot be handled like that, since you are creating a new type.

So what this function instead needs to do is to create a copy of the SymEngine object, so that the caller is always responsible for freeing `cbasic_operand2`. That way, you can create a new SymEngine object Rational and just return it.

Also, why not to use the low level SymEngine API, and just define it as:
```c++
void sympify(VALUE operand2, basic basic_operand2) {
...
    
            num = rb_funcall(operand2, rb_intern("numerator"), 0, NULL);
            den = rb_funcall(operand2, rb_intern("denominator"), 0, NULL);

            num_basic = basic_new_heap();
            den_basic = basic_new_heap();

            GET_SYMINTFROMVAL(num, num_basic);
            GET_SYMINTFROMVAL(den, den_basic); 

            rational_set(basic_operand2, num_basic, den_basic);

            basic_free_heap(num_basic);
            basic_free_heap(den_basic); 
}
```

You can actually simplify this and allocate `num_basic` and `den_basic` on stack, but that's minor.

When you call this, the caller is responsible to call either `basic_init` or `basic_new_heap` on the `basic_operand2` argument, and then it is responsible to free it properly.

Let me know what you think.